### ISSUE TITLE
DDP is not working for me...

### ISSUE THREAD

**Issue by ByungSun12** [PyTorchLightning/pytorch-lightning#3640]
## 🐛 Bug
I run mnist example code in this repo.
When i excuted mnist.py, I could see one gpu's usage is 100% and this process is not finished.
After change ddp to ddp_spawn, It works.
But I want to use ddp!

### To Reproduce

Run [mnist.py](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/pl_examples/basic_examples/mnist.py).

### Environment

script : [mnist.py](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/pl_examples/basic_examples/mnist.py).

You can get the script and run it with:
```
python mnist.py --batch_size 128 --max_epochs 2 --gpus '0,1' --distributed_backend 'ddp' 
```

 - PyTorch Version (e.g., 1.0): 1.6.0+cu101
 - OS (e.g., Linux): Ubuntu 18.04 (Linux)
 - How you installed PyTorch (`conda`, `pip`, source): pip
 - Build command you used (if compiling from source): No
 - Python version: 3.6.9
 - CUDA/cuDNN version: 10.1/7.6.4
 - GPU models and configuration: Geforce 1080
 - Any other relevant information:

### Additional context

![스크린샷 2020-09-24 오후 5 02 53](https://user-images.githubusercontent.com/65884594/94120026-9b242280-fe8a-11ea-920c-b99988cacbdf.png)
![스크린샷 2020-09-24 오후 5 03 06](https://user-images.githubusercontent.com/65884594/94120032-9d867c80-fe8a-11ea-9242-36b0146facfd.png)

**Comment by Borda** [comment]
are you using one or multiple machines for training, by the monitor it seems as a single machine with two GPU cards, right?

**Comment by awaelchli** [comment]
@ByungSun12 You selected 2 epochs, if you look at your console output, these two epochs are already over!
Now the process is at `trainer.test()`, but this is not supported in ddp mode. 
In ddp mode we cannot call both trainer.fit and trainer.test, unfortunately. This only works with spawn.

**Comment by awaelchli** [comment]
cc @williamFalcon

**Comment by ByungSun12** [comment]
@Borda Yes, It's single machine that has two gpus.

@awaelchli Oh.. really? then.. could I test model with only one gpu after training with ddp mode?

**Comment by williamFalcon** [comment]
ummm really?  yeah, let me debug this fit test thing. i thought that would have worked

**Comment by awaelchli** [comment]
In ddp mode, you cannot launch .test after you have trained with fit in ddp mode. 
Obviously, because ddp launches your whole script at once. 
If you really want to run test in the same script, you need to instantiate a new trainer after fit, and this one runs the test without ddp.

**Comment by williamFalcon** [comment]
well... that’s not entirely true. 

if your script defines 

.fit then .test

then ddp would call that script multiple times on parallel. 

fit would execute. the test. 

we jusy need to register that ddp has already been forked so that test knows to not fork off again.

**Comment by awaelchli** [comment]
There is no forking in ddp. This is not about ddp_spawn.

**Comment by ByungSun12** [comment]
I just modify `mnist.py` like below.

![스크린샷 2020-09-28 오전 11 24 59](https://user-images.githubusercontent.com/65884594/94384311-411ea800-017d-11eb-93d3-0de673fcd2a8.png)

After training model, I re-define trainer on trained model with `distributed_backedn='dp'`. 

Actually, This is a temporary expedient not a fundamental solution.
![스크린샷 2020-09-28 오전 11 28 59](https://user-images.githubusercontent.com/65884594/94384544-cefa9300-017d-11eb-811a-84d7cecbdc03.png)

**Comment by awaelchli** [comment]
@ByungSun12 The ddp mode improved alot with the latest relase v1.0. And recently we fixed a probelm with device selection #4297. 
I encourage you to try the master branch or next release on Tuesday. 
If you still run into issues, try to show the issue in the [bug report model](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/pl_examples/bug_report_model.py) and post here. Thanks
